### PR TITLE
[WEB] add UI for transcription

### PR DESCRIPTION
**Pull request by nikvaessen**
This adds a button to invite and kick the transcriber, as well as a label indicating a transcriber is currently in the room. 

It does not currently play a sound when the transcriber joins or leaves, and it nothing is currently displayed on mobile.

### CODE REVIEW

File: react/features/toolbox/components/web/Toolbox.js [hunk 1]
```diff
@@ -894,6 +927,18 @@ class Toolbox extends Component<Props> {
...
+
+    /**
+     * Dispatched an action to toggle transcribing.
+     *
+     * @private
+     * @returns {void}
+     */
+    _onToolbarToggleTranscribing() {
+        this._doToggleTranscribing();
+    }
```
Comment (virtuacoplenny) [inline react/features/toolbox/components/web/Toolbox.js]: Probably want to add some analytics for this event. While adding analytics in jitsi-meet is pretty inconsistent, I think at least toolbox has analytics for all its buttons.

File: react/features/transcribing/components/TranscribingLabel.web.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+     *
+     * @inheritdoc
+     * @param {Object} nextProps - The read-only React {@code Component} props
+     * that this instance will receive.
+     * @returns {void}
+     */
+    componentWillReceiveProps(nextProps) {
+        this.setState({
+            hidden: !nextProps._transcribing
+        });
```
Comment (virtuacoplenny) [inline react/features/transcribing/components/TranscribingLabel.web.js]: Why use state instead of using the transcribing prop directly?

File: react/features/transcribing/middleware.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        transcriberJID,
+        potentialTranscriberJIDs
+    } = store.getState()['features/transcribing'];
+
+    switch (action.type) {
+    case DIAL_TRANSCRIBER:
+        if (!(isDialing || isTranscribing)) {
+            store.dispatch(showPendingTranscribingNotification());
+
+            APP.conference._room.dial(TRANSCRIBER_DIAL_COMMAND).then(
```
Comment (virtuacoplenny) [inline react/features/transcribing/middleware.js]: You have access to the conference through the store if you want to use it. I'd avoid using the global as much as possible because the intent is to get rid of it some day, whenever that day may be.

File: react/features/transcribing/middleware.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                () => {
+                    // do nothing
```
Comment (virtuacoplenny) [inline react/features/transcribing/middleware.js]: Maybe using a .catch instead of .then is more expressive of what you want to do.

File: react/features/transcribing/middleware.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        break;
+    case _TRANSCRIBER_LEFT:
+        store.dispatch(showStoppedTranscribingNotification());
+        break;
+    case HIDDEN_PARTICIPANT_JOINED:
+        if (action.displayName
+                && action.displayName === TRANSCRIBER_DISPLAY_NAME) {
+            store.dispatch(transcriberJoined(action.id));
+        } else {
+            store.dispatch(potentialTranscriberJoined(action.id));
```
Comment (virtuacoplenny) [inline react/features/transcribing/middleware.js]: What conditions cause `potentialTranscriberJoined` to be needed?

File: react/features/transcribing/reducer.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+                isTranscribing: false,
+                transcriberJID: undefined,
+                potentialTranscriberJIDs: []
+            };
+        case _POTENTIAL_TRANSCRIBER_JOINED:
+            return {
+                ...state,
+                potentialTranscriberJIDs:
+                    [ action.transcriberJID ]
+                        .concat(state.potentialTranscriberJIDs)
```
Comment (virtuacoplenny) [inline react/features/transcribing/reducer.js]: What's the scenario in which there can be more than 1 potential subscriber jid? Is it possible for multiple people to start transcribing?

Comment (nikvaessen) [reply react/features/transcribing/reducer.js]: Because Jibri and Jigasi will both use the recorder domain (hidden domain), both Jibri and Jigasi joining the conference will result in a HIDDEN_PARTICIPANT_JOINED event. We check for Jigasi instance by looking for the DisplayName, which should be Transcriber, but the presence causing the event to be emitted does not yet contain the displayname. Thus I store the potential JID's and only start transcribing when the JID is matched to the correct DisplayName

File: react/features/transcribing/components/TranscribingLabel.web.js (hunk 2 above)
Comment (nikvaessen) [reply react/features/transcribing/components/TranscribingLabel.web.js]: I don't understand this question. Do you mean `this.setState(nextProps)`? If so, doesn't  `nextProps` contains the function `t`?

**Comment by nikvaessen** [comment]
How could I test if the analytic event is working?

File: react/features/transcribing/components/TranscribingLabel.web.js (hunk 2 above)
Comment (virtuacoplenny) [reply react/features/transcribing/components/TranscribingLabel.web.js]: Instead of checking this.state.hidden or !this.state.hidden, why not check this.props._transcribing directly?

File: react/features/transcribing/middleware.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    } = store.getState()['features/transcribing'];
+
+    const { conference } = store.getState()['features/base/conference'];
+
+    switch (action.type) {
+    case DIAL_TRANSCRIBER:
+        if (!(isDialing || isTranscribing)) {
+            store.dispatch(showPendingTranscribingNotification());
+
+            conference.room.dial(TRANSCRIBER_DIAL_COMMAND).catch(
```
Comment (virtuacoplenny) [inline react/features/transcribing/middleware.js]: .catch takes in one function, which is the error handler. It's sugar over `.then(undefined, errorCallback)`.

File: react/features/transcribing/middleware.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+                    store.dispatch(dialError());
+                    store.dispatch(hidePendingTranscribingNotification());
+                    store.dispatch(showTranscribingError());
+                }
+            );
+        }
+        break;
+    case STOP_TRANSCRIBING:
+        if (isTranscribing) {
+            APP.conference._room.kickParticipant(transcriberJID);
```
Comment (virtuacoplenny) [inline react/features/transcribing/middleware.js]: You might also want to swap this reference to the global conference for the locally scoped conference.

**Comment by virtuacoplenny** [comment]
> How could I test if the analytic event is working?

When I remember to test it I use the dev proxy server pointed to beta and check if AnalyticsAdapter#sendEvent sends the information I expect.  @bgrozev or @hristoterezov very likely know where the events are going and how to access them.

**Comment by nikvaessen** [comment]
Event on button press worked on Splunk.

File: conference.js [hunk 9]
```diff
@@ -75,6 +75,8 @@ import {
     getAvatarURLByParticipantId,
     getLocalParticipant,
     getParticipantById,
+    hiddenParticipantJoined,
```
Comment (saghul) [inline conference.js]: Drive-by observation: we should not need these. A participant can have a hidden property if that's needed. Or alternatively we could filter transcribers before dispatching the participantJoined action.

Comment (nikvaessen) [reply conference.js]: I disagree; currently a PARTICIPANT_JOINED event triggers whatever logic is required for a new participant, such as displaying it in the filmstrip. All this code should then first check if it's not actually a hidden participant?

File: react/features/transcribing/components/TranscribingLabel.web.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+            return null;
+        }
+
+        return (
+            <Tooltip
+                content = 'This label indicates that there is a transcriber'
+                position = { 'left' }>
+                <CircularLabel
+                    className = 'recording-label'
+                    label = 'TR' />
```
Comment (virtuacoplenny) [inline react/features/transcribing/components/TranscribingLabel.web.js]: It'll be worthwhile to put the label and tooltip content through translation. Check out how other components use `t()` and strings in main.json.

**Comment by nikvaessen** [comment]
Force pushed to the PR... messed up with fixing a merge conflict.

File: css/_font.scss [hunk 11]
```diff
@@ -24,6 +24,10 @@
     -webkit-font-smoothing: antialiased;
     -moz-osx-font-smoothing: grayscale;
 }
+
+.icon-cc:before {
```
Comment (virtuacoplenny) [inline css/_font.scss]: Does this icon really exist? If so where is it used?
